relayer/client.go

```diff
@@ -238,137 +234,182 @@ func CreateClient(ctx context.Context, src, dst *Chain, srcUpdateHeader, dstUpda
+	ctx context.Context,
+	src, dst *Chain,
+	memo string,
+) (err error) {
+	srch, dsth, err := QueryLatestHeights(ctx, src, dst)
 	if err != nil {
-		c.log.Debug(
-			"Failed to update source client",
-			zap.String("src_chain", c.ChainID()),
-			zap.Error(err),
-		)
 		return err
 	}
 
-	dstUpdateMsg, err := dst.ChainProvider.MsgUpdateClient(dst.ClientID(), srcUpdateHeader)
-	if err != nil {
-		dst.log.Debug(
-			"Failed to update destination client",
-			zap.String("dst_chain", dst.ChainID()),
-			zap.Error(err),
-		)
+	var srcMsgUpdateClient, dstMsgUpdateClient provider.RelayerMessage
+	eg, egCtx := errgroup.WithContext(ctx) // New errgroup because previous egCtx is canceled at this point.
+	eg.Go(func() error {
+		var err error
+		srcMsgUpdateClient, err = MsgUpdateClient(egCtx, dst, src, dsth, srch)
+		return err
+	})
+	eg.Go(func() error {
+		var err error
+		dstMsgUpdateClient, err = MsgUpdateClient(egCtx, src, dst, srch, dsth)
+		return err
+	})
+
+	if err = eg.Wait(); err != nil {
 		return err
 	}
 
 	clients := &RelayMsgs{
-		Src: []provider.RelayerMessage{srcUpdateMsg},
-		Dst: []provider.RelayerMessage{dstUpdateMsg},
+		Src: []provider.RelayerMessage{srcMsgUpdateClient},
+		Dst: []provider.RelayerMessage{dstMsgUpdateClient},
 	}
 
 	// Send msgs to both chains
-	result := clients.Send(ctx, c.log, AsRelayMsgSender(c), AsRelayMsgSender(dst), memo)
+	result := clients.Send(ctx, src.log, AsRelayMsgSender(src), AsRelayMsgSender(dst), memo)
 	if err := result.Error(); err != nil {
 		if result.PartiallySent() {
-			c.log.Info(
+			src.log.Info(
 				"Partial success when updating clients",
-				zap.String("src_chain_id", c.ChainID()),
+				zap.String("src_chain_id", src.ChainID()),
 				zap.String("dst_chain_id", dst.ChainID()),
 				zap.Object("send_result", result),
 			)
 		}
 		return err
 	}
 
-	c.log.Info(
+	src.log.Info(
 		"Clients updated",
-		zap.String("src_chain_id", c.ChainID()),
-		zap.String("src_client", c.PathEnd.ClientID),
-		zap.Stringer("src_height", MustGetHeight(srcUpdateHeader.GetHeight())),
-		zap.Uint64("src_revision_height", srcUpdateHeader.GetHeight().GetRevisionHeight()),
+		zap.String("src_chain_id", src.ChainID()),
+		zap.String("src_client", src.PathEnd.ClientID),
 
 		zap.String("dst_chain_id", dst.ChainID()),
 		zap.String("dst_client", dst.PathEnd.ClientID),
-		zap.Stringer("dst_height", MustGetHeight(dstUpdateHeader.GetHeight())),
-		zap.Uint64("dst_revision_height", dstUpdateHeader.GetHeight().GetRevisionHeight()),
 	)
 
 	return nil
 }
 
-// UpgradeClients upgrades the client on src after dst chain has undergone an upgrade.
-func (c *Chain) UpgradeClients(ctx context.Context, dst *Chain, height int64, memo string) error {
-	dstHeader, err := dst.ChainProvider.GetLightSignedHeaderAtHeight(ctx, height)
+// UpgradeClients upgrades the client on dst after src chain has undergone an upgrade.
+func UpgradeClient(
+	ctx context.Context,
+	src, dst *Chain,
+	height int64,
+	memo string,
+) (err error) {
```

Comment:
Piggybacking off Mark.
I believe the named return is unnecessary here.

Reply:
done